include/oneapi/dpl/pstl/tuple_impl.h

```diff
@@ -289,6 +306,82 @@ struct __copy_assignable_holder<_Tp, false> : oneapi::dpl::__internal::__value_h
     operator=(__copy_assignable_holder&& other) = default;
 };
 
+template <typename _Tuple1, typename _Tuple2, int I = 0>
+constexpr bool
+__equal(_Tuple1&& __lhs, _Tuple2&& __rhs)
+{
+    static_assert(std::tuple_size_v<std::decay_t<_Tuple1>> == std::tuple_size_v<std::decay_t<_Tuple2>>,
+                  "Tuples must have the same size to be equality compared");
+    if constexpr (I < std::tuple_size_v<std::decay_t<_Tuple1>>)
+    {
+        return std::get<I>(__lhs) == std::get<I>(__rhs) &&
+               oneapi::dpl::__internal::__equal<_Tuple1, _Tuple2, I + 1>(std::forward<_Tuple1>(__lhs),
+                                                                         std::forward<_Tuple2>(__rhs));
+    }
+    else
+    {
+        return true;
+    }
+}
+template <typename _Tuple1, typename _Tuple2, int I = 0>
+constexpr bool
+__less(_Tuple1&& __lhs, _Tuple2&& __rhs)
+{
+    static_assert(std::tuple_size_v<std::decay_t<_Tuple1>> == std::tuple_size_v<std::decay_t<_Tuple2>>,
+                  "Tuples must have the same size to be compared");
+    if constexpr (I < std::tuple_size_v<std::decay_t<_Tuple1>>)
+    {
+        return std::get<I>(__lhs) < std::get<I>(__rhs) ||
+               (!(std::get<I>(__rhs) < std::get<I>(__lhs)) &&
+                oneapi::dpl::__internal::__less<_Tuple1, _Tuple2, I + 1>(std::forward<_Tuple1>(__lhs),
+                                                                         std::forward<_Tuple2>(__rhs)));
+    }
+    else
+    {
+        return false;
+    }
+}
+
+template <typename T>
+struct __is_internal_tuple : std::false_type
+{
+};
+
+template <typename... Ts>
+struct __is_internal_tuple<oneapi::dpl::__internal::tuple<Ts...>> : std::true_type
+{
+};
+
+template <typename T>
+inline constexpr bool __is_internal_tuple_v = __is_internal_tuple<T>::value;
+
+template <typename T>
+struct __is_std_tuple : std::false_type
+{
+};
+
+template <typename... Ts>
+struct __is_std_tuple<std::tuple<Ts...>> : std::true_type
+{
+};
+
+template <typename T>
+inline constexpr bool __is_std_tuple_v = __is_std_tuple<T>::value;
+
+// Either LHS = this internal tuple type + RHS = any compatible tuple type
+// or     LHS = std::tuple               + RHS = this internal tuple type
+// Note: We do not allow LHS to be "any compatible type" with a RHS as this internal tuple type because then we would
+//       have ambiguity comaparing two different compatible internal tuple types, because both instantiations would
+//       create the necessary comparison. As written, the type instatation of the type in the LHS is always
+//       responsible for creating this comparison operator.
+template <typename _ThisTuple, typename _U, typename _V>
+inline constexpr bool __enable_comparison_op_v =
```

Comment:
Will we have any benefits here if use [std::disjunction](https://en.cppreference.com/w/cpp/types/disjunction) and then define `__enable_comparison_op_v` as `...::value` ?

Reply:
I suppose it would allow us to omit defining `__is_std_tuple_v` and `__is_internal_tuple_v`.  I'm not aware of any other real advantages here.  
I think we may get some very slight compile time improvement, because I think that more might be able to be "short circuited" here and not evaluated, but I can't imagine that is significant.

I prefer the current formulation because I find it makes the most clear structure for the logic below.

For a similar reason, I might be in favor of removing `oneapi::dpl::__internal::` as @akukanov  mentions.  However, I think it has been our most recent convention to use the fully qualified name at all times.

Reply:
ok.